File: app/src/main/java/com/ivanovsky/passnotes/ui/groups/GroupsPresenter.java
package com.ivanovsky.passnotes.ui.groups;

import android.content.Context;

import com.ivanovsky.passnotes.App;
import com.ivanovsky.passnotes.data.ObserverBus;
import com.ivanovsky.passnotes.data.safedb.EncryptedDatabaseProvider;
import com.ivanovsky.passnotes.data.safedb.GroupRepository;
import com.ivanovsky.passnotes.data.safedb.model.Group;
import com.ivanovsky.passnotes.ui.core.FragmentState;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class GroupsPresenter implements
		GroupsContract.Presenter,
		ObserverBus.GroupDataSetObserver {

	@Inject
	EncryptedDatabaseProvider dbProvider;

	@Inject
	ObserverBus observerBus;

	private final Context context;
	private final GroupsContract.View view;
	private final CompositeDisposable disposables;

	GroupsPresenter(Context context, GroupsContract.View view) {
		App.getDaggerComponent().inject(this);
		this.context = context;
		this.view = view;
		this.disposables = new CompositeDisposable();
	}

	@Override
	public void start() {
		view.setState(FragmentState.LOADING);
		observerBus.register(this);
		loadData();
	}

	@Override
	public void stop() {
		observerBus.unregister(this);
		disposables.clear();
	}

	@Override
	public void loadData() {
		if (dbProvider.isOpened()) {
			GroupRepository repository = dbProvider.getDatabase().getGroupRepository();

			Disposable disposable = repository.getAllGroup()
					.subscribeOn(Schedulers.newThread())
					.observeOn(AndroidSchedulers.mainThread())
					.subscribe(this::onGroupsLoaded);
			disposables.add(disposable);
		} else {
			view.showUnlockScreenAndFinish();
		}
	}

	private void onGroupsLoaded(List<Group> groups) {
		if (groups.size() != 0) {
			view.showGroups(groups);
		} else {
			view.showNoItems();
		}
	}

	@Override
	public void onGroupDataSetChanged() {
		loadData();
	}
}

File: app/src/main/kotlin/com/ivanovsky/passnotes/ui/groups/GroupsPresenter.kt
package com.ivanovsky.passnotes.ui.groups

import android.content.Context
import com.ivanovsky.passnotes.data.ObserverBus
import com.ivanovsky.passnotes.data.entity.Group
import com.ivanovsky.passnotes.domain.interactor.groups.GroupsInteractor
import com.ivanovsky.passnotes.injection.Injector
import com.ivanovsky.passnotes.presentation.core.FragmentState
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer
import javax.inject.Inject

class GroupsPresenter(val context: Context, val view: GroupsContract.View) :
		GroupsContract.Presenter,
		ObserverBus.GroupDataSetObserver {

	@Inject
	lateinit var interactor: GroupsInteractor

	@Inject
	lateinit var observerBus: ObserverBus

	private val disposables: CompositeDisposable

	override fun start() {
		view.setState(FragmentState.LOADING)
		observerBus.register(this)
		loadData()
	}

	init {
		Injector.getInstance().encryptedDatabaseComponent.inject(this)
		disposables = CompositeDisposable()
	}

	override fun stop() {
		observerBus.unregister(this)
		disposables.clear()
	}

	override fun loadData() {
		val disposable = interactor.getAllGroupsWithNoteCount()
				.subscribe(Consumer { onGroupsLoaded(it) })

		disposables.add(disposable)
	}

	private fun onGroupsLoaded(groupsAndCounts: List<Pair<Group, Int>>) {
		if (groupsAndCounts.isNotEmpty()) {
			view.showGroups(groupsAndCounts)
		} else {
			view.showNoItems()
		}
	}

	override fun onGroupDataSetChanged() {
		loadData()
	}

	override fun onGroupClicked(group: Group) {
		view.showNotesScreen(group)
	}
}